sacc2/sacc.py

```diff
@@ -0,0 +1,1026 @@
+            These should either be already set up using the add_tracer method,
+            or you could set tracers_laster=True if you want to add them later.
+            e.g. for 2pt measurements the tracers are the names of the two n(z)
+            samples
+
+        value: float
+            A single value for the data point
+
+        tracers_later: bool
+            If True, do not complain if the tracers are not know already
+
+        **tags:
+            Tags to apply to this data point.
+            Tags can be any arbitrary metadata that you might want later,
+            For 2pt data the tag would include an angle theta or ell.
+
+        Returns
+        -------
+        None
+        """
+        if self.covariance is not None:
+            raise ValueError("You cannot add a data point after setting the covariance")
+        tracers = tuple(tracers)
+        for tracer in tracers:
+            if (tracer not in self.tracers) and (not tracers_later):
+                raise ValueError(f"Tracer named '{tracer}' is not in the known list of tracers."
+                    "Either put it in before adding data points or set tracers_later=True")
+        d = DataPoint(data_type, tracers, value, **tags)
+        self.data.append(d)
+
+    def add_covariance(self, covariance):
+        """
+        Once you have finished adding data points, add a covariance
+        for the entire set.
+
+        Parameters
+        ----------
+        covariance: array or list
+            2x2 numpy array containing the covariance of the added data points
+            OR a list of blocks
+
+        Returns
+        -------
+        None
+        """
+        if isinstance(covariance, BaseCovariance):
+            self.covariance = covariance
+        else:
+            self.covariance = BaseCovariance.make(covariance, len(self))
+
+
+    def cut(self, mask):
+        """
+        Remove data points and corresponding covariance elements following a mask, in-place.
+
+        You can find indices (e.g. matching some tag) to remove using the indices method.
+
+        Parameters
+        ----------
+        mask: array or list
+            Mask must be either a boolean array or a list/array of integer indices to remove.
+            if boolean then True means to cut data point and False means to keep it
+            if indices then values indicate data points to cut out
+        """
+        mask = np.array(mask)
+
+        # Convert integer masks to booleans
+        if mask.dtype != np.bool:
+            if not mask.dtype in [np.int8, np.int16, np.int32, np.int64]:
+                raise ValueError("Wrong mask type")
+            m = np.zeros(len(self), dtype=bool)
+            for i in mask:
+                m[i] = True
+            mask = m
+            
+        if not len(mask)==len(self):
+            raise ValueError("Mask passed in is wrong size")
+
+        self.data = [d for i,d in enumerate(self.data) if not mask[i]]
```

Comment:
I might worry that this kind of loops might make things slow (as opposed to a simple evaluation of an ndarray on `mask`). Has this been benchmarked?

Reply:
There are other instances of this (e.g. in `reorder`)

Reply:
Running reorder on 2000 data points takes 4ms.  It should scale as O(n log(n)).

Reply:
The sort, I mean.

Reply:
ok, no problem then!